src/top/data/bbox_reg_util.py

```diff
@@ -86,72 +89,85 @@ class CropObject(object):
+        
+        vis_img = th.tensor([])
+        vis_crop_img = th.tensor([])
+        vis_point_2d = th.tensor([])
+        vis_trans = th.tensor([])
+        vis_orient = th.tensor([])
+        vis_scale = th.tensor([])
         
-        orientation = np.split(orientation, num_object)
-        orientation = [rotation.reshape(3,3) for rotation in orientation]
-
-        scale = np.split(scale, num_object)
-        scale = [scales.reshape(-1,3) for scales in scale]
-
-        translation = np.split(translation, num_object)
-        translation = [translations.reshape(-1,3) for translations in translation]
-
-        crop_img = []
-        quaternions = []
-        _scale = []
-        _translation = []
-        for object_id in range(num_object):       
-            # NOTE(Jiyong): np.split() leaves an empty array at the end of the list.
-            if keypoints_2d[object_id].size == 0:
-                break
-
-            x_min, y_min, _ = np.min(keypoints_2d[object_id], axis=0)
-            x_max, y_max, _ = np.max(keypoints_2d[object_id], axis=0)
-
-            # NOTE(Jiyong): TypeError: Expected cv::UMat for argument 'src'
-            # -> cv2.Umat() is functionally equivalent to np.float32() & (H,W,C)
-            crop_tmp = np.float32(image[:, y_min:y_max, x_min:x_max])
-            crop_tmp = np.transpose(crop_tmp, (1,2,0))
-            try:
-                crop_tmp = cv2.resize(crop_tmp, dsize=self.opts.crop_img_size)
-            except Exception as e:
-                print(crop_tmp.shape)
-                raise
-
-            crop_tmp = np.transpose(crop_tmp, (2,0,1))
-            crop_img.append(crop_tmp)
-
-            # For quaternions regression
-            r = R.from_matrix(orientation[object_id])
-            quaternions.append(r.as_quat())
-
-            _scale.append(scale[object_id])
-            _translation.append(translation[object_id])
+        for obj in range(num_object):
+            # NOTE(Jiyion): If visiblity is false, cropping that object is impossible
+            if not visibility[obj]:
+                continue
 
+            crop_tmp = image[:, int(point_min[obj][1]):int(point_max[obj][1]), int(point_min[obj][0]):int(point_max[obj][0])]
+            crop_tmp = resize(crop_tmp, size=self.opts.crop_img_size)
+
+            vis_img = th.cat((vis_img, image))
+            vis_crop_img = th.cat((vis_crop_img, crop_tmp))
+            vis_point_2d = th.cat((vis_point_2d, keypoints_2d[obj]))
+            vis_trans = th.cat((vis_trans, translation[obj]))
+            vis_orient = th.cat((vis_orient, quaternions[obj]))
+            vis_scale = th.cat((vis_scale, scale[obj]))
+            
+        # shallow copy
         outputs = inputs.copy()
-        outputs['crop_img'] = np.stack(crop_img, axis=0)
-        outputs[Schema.TRANSLATION] = _translation
-        outputs[Schema.SCALE] = _scale
-        outputs[Schema.ORIENTATION] = quaternions
+        outputs[Schema.IMAGE] = vis_img.reshape(-1, c, w, h)
```

Comment:
I think we should just get rid of `outputs[Schema.IMAGE]`, i.e. `outputs.pop(Schema.IMAGE)` Unless you think holding onto this is useful. The reason to avoid this would be to free unused memory ASAP.

Reply:
Yeah, I agree. I will get rid of `outputs[Schema.IMAGE]`.